homeassistant/components/reolink/config_flow.py

```diff
@@ -96,7 +96,37 @@ async def async_step_reauth_confirm(
     async def async_step_dhcp(self, discovery_info: dhcp.DhcpServiceInfo) -> FlowResult:
         """Handle discovery via dhcp."""
         mac_address = format_mac(discovery_info.macaddress)
-        await self.async_set_unique_id(mac_address)
+        existing_entry = await self.async_set_unique_id(mac_address)
+        if (
+            existing_entry
+            and CONF_PASSWORD in existing_entry.data
+            and existing_entry.data[CONF_HOST] != discovery_info.ip
+        ):
+            # check if the camera is reachable at the new IP
+            host = ReolinkHost(self.hass, existing_entry.data, existing_entry.options)
+            try:
+                host.api.get_state("GetLocalLink")
+                host.api.logout()
+            except ReolinkError as err:
+                _LOGGER.debug(
+                    "Reolink DHCP reported new IP '%s', "
+                    "but got error '%s' trying to connect, so sticking to IP '%s'",
+                    discovery_info.ip,
+                    str(err),
+                    existing_entry.data[CONF_HOST],
+                )
+                raise AbortFlow("already_configured") from err
+            if format_mac(host.api.mac_address) != mac_address:
+                _LOGGER.debug(
+                    "Reolink mac address '%s' at new IP '%s' from DHCP, "
+                    "does not match mac '%s' of config entry, so sticking to IP '%s'",
+                    format_mac(host.api.mac_address),
+                    discovery_info.ip,
+                    mac_address,
```

Comment:
This is the mac address of the discovery info. That's not for sure the same as the mac address of the config entry, right?

Reply:
It is, because in the line above this mac_address is used to get the config entry, if it would not match the config entry existing_entry would be None and the code in the `if `statement would be skipped.
`existing_entry = await self.async_set_unique_id(mac_address)`

Reply:
At least that is what I think is supposed to happen.....

Reply:
Ah, right, yes.